flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkTaskWriter.java

```diff
@@ -0,0 +1,212 @@
+  private final PartitioningWriter<PositionDelete<RowData>, DeleteWriteResult> positionDeleteWriter;
+
+  private final PartitionSpec spec;
+  private final FileIO io;
+  private final boolean upsert;
+  private final Function<RowData, StructLike> partitioner;
+  private final Function<RowData, StructLike> equalityKeyer;
+
+  private StructLike partition;
+  private StructLike equalityKey;
+  private final PositionDelete<RowData> positionDelete;
+  private final Map<StructLike, PathOffset> insertedRowMap; // map row Key to PathOffset
+
+  protected FlinkTaskWriter(
+      PartitioningWriterFactory<RowData> partitioningWriterFactory,
+      Function<RowData, StructLike> partitioner,
+      Schema schema, PartitionSpec spec, Set<Integer> equalityFieldIds,
+      FileIO io, RowType flinkSchema, boolean upsert) {
+    this.insertWriter = partitioningWriterFactory.newDataWriter();
+    this.equalityDeleteWriter = partitioningWriterFactory.newEqualityDeleteWriter();
+    this.positionDeleteWriter = partitioningWriterFactory.newPositionDeleteWriter();
+
+    this.spec = spec;
+    this.io = io;
+    this.upsert = upsert;
+
+    this.insertedRowMap = StructLikeMap.create(TypeUtil.select(schema.asStruct(), equalityFieldIds));
+    this.positionDelete = PositionDelete.create();
+    this.partitioner = partitioner;
+    this.equalityKeyer = equalityKeyerFor(equalityFieldIds, schema, flinkSchema);
+  }
+
+  @Override
+  @SuppressWarnings("DuplicateBranchesInSwitch")
+  public void write(RowData row) throws IOException {
+    partition = partitioner.apply(row);
+    equalityKey = equalityKeyer.apply(row);
+
+    switch (row.getRowKind()) {
+      case INSERT:
+        if (upsert) {
+          delete(row);
+        }
+        insert(row);
+        break;
+      case UPDATE_BEFORE:
+        if (upsert) {
+          break;  // UPDATE_BEFORE is not necessary for UPDATE, ignore to prevent deleting twice
+        }
+        delete(row);
+        break;
+      case UPDATE_AFTER:
+        if (upsert) {
+          delete(row);
+        }
+        insert(row);
+        break;
+      case DELETE:
+        delete(row);
+        break;
+      default:
+        throw new UnsupportedOperationException("Unknown row kind: " + row.getRowKind());
+    }
+  }
+
+  protected boolean internalPosDelete(StructLike key) {
+    PathOffset previous = insertedRowMap.remove(key);
+    if (previous != null) {
+      positionDeleteWriter.write(previous.setTo(positionDelete), spec, partition);
+      return true;
+    }
+    return false;
+  }
+
+  protected void insert(RowData row) throws IOException {
+    StructLike copiedKey = StructCopy.copy(equalityKey);
+
+    internalPosDelete(copiedKey);
+    PathOffset pathOffset = insertWriter.write(row, spec, partition);
+    insertedRowMap.put(copiedKey, pathOffset);
+  }
+
+  protected void delete(RowData row) throws IOException {
+    StructLike copiedKey = StructCopy.copy(equalityKey);
```

Comment:
In this `delete` method, we don't need to deep copy a new `equalityKey` because we won't hold any reference in the global `insertedRowMap` or other shared memory.